### PR TITLE
Shopping cart ryan us23 24

### PR DESCRIPTION
**Pull request by cmpprg**
- Refactor some instance methods on cart PORO
- Add patch route to cart controller
- Add 'add_or_sub' action to cart controller
- Add #max_quantity?, #decrement, #remove, #quantity_zero? to cart PORO
- Add ability to increment and decrement to cart show view

### CODE REVIEW

File: app/controllers/cart_controller.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 class CartController < ApplicationController
   def add_item
     item = Item.find(params[:item_id])
-    cart.add_item(item.id.to_s)
+    cart.add_item(params[:item_id])
```
Comment (tony-gomes) [inline app/controllers/cart_controller.rb:4]: much cleaner

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -9,7 +9,7 @@ class ApplicationController < ActionController::Base
                 :visitor?
 
   def cart
-    @cart ||= Cart.new(session[:cart] ||= Hash.new(0))
+    @cart ||= Cart.new(session[:cart] ||= {} )
```
Comment (tylerpporter) [inline app/controllers/application_controller.rb:12]: Great catch, this change is great.

File: app/controllers/cart_controller.rb [hunk 3]
```diff
@@ -21,13 +21,19 @@ def remove_item
...
-  #   if params[:increment_decrement] == "increment"
-  #     cart.add_quantity(params[:item_id]) unless cart.limit_reached?(params[:item_id])
-  #   elsif params[:increment_decrement] == "decrement"
-  #     cart.subtract_quantity(params[:item_id])
-  #     return remove_item if cart.quantity_zero?(params[:item_id])
-  #   end
-  #   redirect_to "/cart"
-  # end
+  def add_or_sub
+    params[:add?] ? increment(params[:item_id]) : decrement(params[:item_id])
```
Comment (tony-gomes) [inline app/controllers/cart_controller.rb:25]: great use of a ternary

**Review by Shock253 — APPROVED** [review]

File: app/models/cart.rb [hunk 4]
```diff
@@ -3,23 +3,22 @@ class Cart
 
   def initialize(contents)
     @contents = contents
+    @contents.default = 0
```
Comment (tony-gomes) [inline app/models/cart.rb:6]: Rubyist solutons rule

**Review by tylerpporter — APPROVED** [review]
Remember to test the cart methods on the model level.

File: spec/features/cart/show_spec.rb [hunk 5]
```diff
@@ -55,6 +55,71 @@
...
+        end
+
+        within("#cart-item-#{@tire.id}") do
+          15.times do
+            click_link("+1")
+          end
+        end
+
+        within("#cart-item-#{@tire.id}") do
+          expect(page).to have_content("12")
```
Comment (tony-gomes) [inline spec/features/cart/show_spec.rb:83]: great testing...

**Review by tony-gomes — APPROVED** [review]
Looks great...

**Pull request merged**